### PR TITLE
[SYCL-MLIR] Add `noalias` attribute for `restrict` function arguments

### PR DESCRIPTION
**Pull request by whitneywhtsang**

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc [hunk 1]
```diff
@@ -409,6 +409,28 @@ const CodeGenOptions &CodeGenTypes::getCodeGenOpts() const {
   return CGM.getCodeGenOpts();
 }
 
+// Gets the declared type of a function parameter when given the index of the
+// parameter and the decayed type of the parameter.
+static QualType getDeclArgTy(const FunctionDecl &FD, int32_t ArgNo,
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.cc:414]: Changed lambda to a static function for reuse.

File: polygeist/tools/cgeist/Test/Verification/sycl/restrict.cpp [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// RUN: clang++ -fsycl -fsycl-device-only -O0 -w -emit-mlir %s -o - | FileCheck %s --check-prefix=CHECK-MLIR
+// RUN: clang++ -fsycl -fsycl-device-only -O0 -w -S -emit-llvm -fsycl-targets=spir64-unknown-unknown-syclmlir %s -o - | FileCheck %s --check-prefix=CHECK-LLVM
+
+// CHECK-MLIR: func.func @func(%arg0: memref<?xi32, 4> {{{.*}}llvm.noalias{{.*}}}, %arg1: memref<?xi32, 4> {{{.*}}llvm.noalias{{.*}}}) 
+// CHECK-LLVM: define spir_func {{.*}}i32 @func(i32 addrspace(4)* noalias {{.*}}%0, i32 addrspace(4)* noalias {{.*}}%1)
+
+#include <sycl/sycl.hpp>
+
+extern "C" SYCL_EXTERNAL int func(int * __restrict__ a, int * __restrict__ b) {
```
Comment (etiotto) [inline polygeist/tools/cgeist/Test/Verification/sycl/restrict.cpp:9]: Let's also test restrict qualifiers on:
   - a struct argument
   - a vector
   - a member function of a class (which receives pointers to class)

Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Test/Verification/sycl/restrict.cpp:9]: `error: pointer to function type cannot have '__restrict' qualifier`
Do you know how to add restrict to member function or any function pointer?

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc [hunk 3]
```diff
@@ -409,6 +409,28 @@ const CodeGenOptions &CodeGenTypes::getCodeGenOpts() const {
...
+                             QualType ABIArgTy) {
+  const bool IsMethodDecl = isa<CXXMethodDecl>(FD);
+  const bool IsMethodInstance =
+      IsMethodDecl && cast<CXXMethodDecl>(FD).isInstance();
+  if (IsMethodInstance) // account for the fact the 'this' type is not
+    ArgNo--;            // present in the function declaration.
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.cc:420]:
```suggestion
  const auto MethodDecl = dyn_cast<CXXMethodDecl>(FD);
  const bool IsMethodInstance = MethodDecl && MethodDecl->isInstance();
  if (IsMethodInstance) // account for the fact the 'this' type is not
    ArgNo--;            // present in the function declaration.
```
NIT

**Review by victor-eds — APPROVED** [review]

File: polygeist/tools/cgeist/Test/Verification/sycl/restrict.cpp (hunk 2 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Test/Verification/sycl/restrict.cpp]: Try:
```
class B { ... };
class A {
   void foo(A * restrict a, B * restrict b) const; 
};
```

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc (hunk 3 above)
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/CodeGenTypes.cc]: changed to `dyn_cast`.

**Review by etiotto — APPROVED** [review]
LGTM

**Pull request merged**